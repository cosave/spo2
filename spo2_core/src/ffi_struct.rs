use syn::{Field, ItemStruct, __private::ToTokens};

const BASE_TYPES: [&str; 11] = [
    "bool", "i8", "i16", "i32", "i64", "u8", "u16", "u32", "u64", "f32", "f64",
];

/// A helper struct used to generate extra code for Rust structs that are
/// intended for use in FFI code.
/// Ideally such structs will have `#[derive(FfiStruct)]` attribute.
///
/// This is used in [spo2_derive] crate to create additional code for common
/// FFI functionality like dropping the struct.
/// This is used in [spo2] for creating foreign language code for
/// the decorated struct.
#[cfg_attr(feature = "develop", derive(Debug))]
pub struct FfiStruct {
    struct_def: ItemStruct,
}

impl FfiStruct {
    /// Creates a new instance of the `FfiStruct` from a struct definition.
    pub fn new(struct_def: ItemStruct) -> Self {
        Self { struct_def }
    }

    /// Get a list of fields in this struct that can safely pass through FFI
    /// without needing special conversion and memory/pointer management.
    /// At this moment only base types are considered safe.
    fn ffi_safe_fields(&self) -> Vec<&Field> {
        self.struct_def
            .fields
            .iter()
            .filter_map(|field| {
                if BASE_TYPES.contains(&field.ty.to_token_stream().to_string().as_str()) {
                    Some(field)
                } else {
                    None
                }
            })
            .collect()
    }

    /// Check if this struct can be safely copied over FFI.
    /// A struct can be safely copied over FFI if it contains only safe fields.
    pub fn is_copy_safe(&self) -> bool {
        self.ffi_safe_fields().len() == self.struct_def.fields.len()
    }
}

#[cfg(test)]
mod tests {
    use syn::{parse2, parse_quote, ItemStruct};

    use crate::FfiStruct;

    #[test]
    fn test_is_copy_safe() {
        let test_cases = [
            (parse_quote! { struct MyStruct {} }, true, "Empty struct."),
            (
                parse_quote! { struct MyStruct { field1: i32, field2: f64, field3: bool } },
                true,
                "Struct with only base types.",
            ),
            (
                parse_quote! { struct MyStruct { field1: *mut c_char, field2: f64, field3: bool } },
                // *const c_char is copy type as it is just pointer's address.
                // TODO: test for pass after we parse raw pointers.
                false,
                "Struct with string pointer.",
            ),
            (
                parse_quote! { struct MyStruct { field1: [u8; 16], field2: f64, field3: bool } },
                // An array is safe as long as the contents are safe.
                // TODO: test for pass after we parse array.
                false,
                "Struct with array or bytes.",
            ),
            (
                parse_quote! { struct MyStruct { field1: Uuid, field2: f64, field3: bool } },
                // A 3rd party crate type such as Uuid is safe as long as it has
                // repr(C) or repr(transparent) attribute and implement Copy trait.
                // TODO: test for pass after we parse 3rd party crates.
                false,
                "Struct with Uuid.",
            ),
            (
                parse_quote! { struct MyStruct { field1: i32, field2: f64, field3: Vec<u8> } },
                false,
                "Struct with a vec.",
            ),
            (
                parse_quote! { struct MyStruct { field1: MyCustomType, field2: MyOtherCustomType } },
                false,
                "Struct with a custom type.",
            ),
            (
                parse_quote! { struct MyStruct { field1: i32, field2: MyNestedStruct } },
                false,
                "Struct with a nested struct.",
            ),
        ];

        for (input, expected, msg) in test_cases.into_iter() {
            let input_struct =
                parse2::<ItemStruct>(input).expect(&format!("Unable to parse: {}", msg));
            let ffi_struct = FfiStruct::new(input_struct);
            assert_eq!(
                ffi_struct.is_copy_safe(),
                expected,
                "Test case {} failed: expected={}, got={}",
                msg,
                expected,
                ffi_struct.is_copy_safe()
            );
        }
    }
}
