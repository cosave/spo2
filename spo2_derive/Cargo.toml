[package]
name = "spo2_derive"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1"
quote = "1"
spo2_core = { path = "../spo2_core" }
syn = { version = "2", features = ["full"] }

[dev-dependencies]
trybuild = "1"
