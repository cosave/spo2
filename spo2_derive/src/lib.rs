use proc_macro::TokenStream;
use quote::quote;
use spo2_core::FfiStruct;
use syn::{parse2, ItemStruct};

#[proc_macro_attribute]
pub fn ffi_struct(_args: TokenStream, input: TokenStream) -> TokenStream {
    let input_struct = FfiStruct::try_from(proc_macro2::TokenStream::from(input.clone()))
        .expect("Unable to parse as Struct");
    if input_struct.is_copy_safe() {
        input
    } else {
        quote! {
             compile_error!("Only copy type structs are supported");
        }
        .into()
    }
}

#[proc_macro_attribute]
pub fn ffi_impl(_args: TokenStream, input: TokenStream) -> TokenStream {
    let input_struct =
        parse2::<ItemStruct>(input.clone().into()).expect(&format!("Unable to parse as Struct"));
    let input_struct = FfiStruct::new(input_struct);
    if input_struct.is_copy_safe() {
        input
    } else {
        quote! {
             compile_error!("Only copy type structs are supported");
        }
        .into()
    }
}

#[cfg(test)]
mod tests {}
