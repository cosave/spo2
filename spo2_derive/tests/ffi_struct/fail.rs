use std::ffi::c_char;

use spo2_derive::ffi_struct;

#[ffi_struct]
struct CPointerStruct {
    field1: *mut c_char,
    field2: f64,
    field3: bool,
}

pub fn main() {}
