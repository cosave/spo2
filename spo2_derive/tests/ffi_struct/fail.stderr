error: Only copy type structs are supported
 --> tests/ffi_struct/fail.rs:5:1
  |
5 | #[ffi_struct]
  | ^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `ffi_struct` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused import: `std::ffi::c_char`
 --> tests/ffi_struct/fail.rs:1:5
  |
1 | use std::ffi::c_char;
  |     ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
